packages/flutter/lib/src/material/button.dart

```diff
@@ -255,12 +259,14 @@ class _MaterialButtonState extends State<MaterialButton> {
     final ButtonTheme buttonTheme = ButtonTheme.of(context);
     final double height = config.height ?? buttonTheme.height;
     final int elevation = (_highlight ? config.highlightElevation : config.elevation) ?? 0;
+    final bool hasColorOrElevation = (config.color != null || elevation > 0);
     Widget contents = new IconTheme.merge(
       context: context,
       data: new IconThemeData(
         color: textColor
       ),
       child: new InkWell(
+        borderRadius: hasColorOrElevation ? BorderRadius.zero : kMaterialButtonDefaultBorderRadius,
```

Comment:
Actually, this will still result in double-clipping. It'll just be that one clip (from the ink response) is a rectangle, and the other clip (from the material painting the ink response) will happen to be rounded at around the same size.